src/software/util/design_patterns/constructor_arg_generic_factory.tpp

```diff
@@ -0,0 +1,111 @@
+#include "software/util/design_patterns/constructor_arg_generic_factory.h"
+
+template <class IndexType, class TypeToCreate>
+ConstructorArgGenericRegistry<IndexType, TypeToCreate>&
+ConstructorArgGenericFactory<IndexType, TypeToCreate>::getMutableRegistry()
+{
+    static ConstructorArgGenericRegistry<IndexType, TypeToCreate> instance;
+    return instance;
+}
+
+template <class IndexType, class TypeToCreate>
+const ConstructorArgGenericRegistry<IndexType, TypeToCreate>&
+ConstructorArgGenericFactory<IndexType, TypeToCreate>::getRegistry()
+{
+    return ConstructorArgGenericFactory::getMutableRegistry();
+}
+
+template <class IndexType, class TypeToCreate>
+std::vector<std::string> ConstructorArgGenericFactory<IndexType, TypeToCreate>::getRegisteredNames()
+{
+    std::vector<std::string> names;
+
+    for (auto iter = ConstructorArgGenericFactory<IndexType, TypeToCreate>::getRegistry().begin();
+         iter != ConstructorArgGenericFactory<IndexType, TypeToCreate>::getRegistry().end(); iter++)
```

Comment:
Could we replace this with `for (auto i : ConstructorArgGenericFactory<IndexType, TypeToCreate>::getRegistry())`?